File: app/src/main/java/com/adobe/marketing/mobile/internal/eventhub/Extensions.java
package com.adobe.marketing.mobile.internal.eventhub;

import kotlin.jvm.functions.Function1;

class ExtensionA extends Extension {

    ExtensionA(final ExtensionApi extensionApi) {
        super(extensionApi);
    }
    @Override
    protected void onRegistered() {
        getApi().registerEventListener("","", this::processEvent);
        Log.print("ExtensionA - onRegistered is done.");
    }
    void processEvent(final Event event){
        Log.print("ExtensionA received event: " + event.getNumber());
    }
}

class ExtensionB extends Extension {

    ExtensionB(final ExtensionApi extensionApi) {
        super(extensionApi);
    }
    @Override
    protected void onRegistered() {
        getApi().registerEventListener("","", this::processEvent);
        try {
            // IO operations: load local cached data
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        Log.print("ExtensionB - onRegistered (block 1 second) is done.");
    }
    void processEvent(final Event event){
        Log.print("ExtensionB received event: " + event.getNumber());
        if (event.getNumber() == 2){
            getApi().stopEvents();
            Log.print("ExtensionB stopEvents at event 2 -------");
            BlockingOperation.sendRequest(event.getNumber(), hit -> {
                Log.print("ExtensionB: request [callback] is triggered: " + event.getNumber());
                getApi().startEvents();
                Log.print("ExtensionB startEvents at event 2 -------");
                return null;
            });
            return;
        }
        if (event.getNumber() == 3){
            Log.print("ExtensionB - 3 ");
        }
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        Log.print("ExtensionB processed event: " + event.getNumber());
    }

}

class ExtensionC extends Extension {

    private int counter = 0;
    ExtensionC(final ExtensionApi extensionApi) {
        super(extensionApi);
    }
    @Override
    protected void onRegistered() {
        getApi().registerEventListener("","", this::processEvent);
        Log.print("ExtensionC - onRegistered is done.");
    }
    void processEvent(final Event event){
        Log.print("ExtensionC received event: " + event.getNumber());
    }
    @Override
    public boolean readyForEvent(final Event event) {
        return ++this.counter > 4;
    }

}


class BlockingOperation{
    static void saveHit(final int hit){
        try {
            Thread.sleep(1000);
            Log.print("Hit saved: " + hit);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    static void sendRequest(final int hit, final Function1<Integer, Void> callback){
        BlockingOperation.saveHit(hit);
        new Thread(() -> {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            callback.invoke(hit);
        }).start();
    }
}

File: app/src/main/java/com/adobe/marketing/mobile/internal/eventhub/containerv2/Extensions.kt
package com.adobe.marketing.mobile.internal.eventhub.containerv2

import com.adobe.marketing.mobile.internal.eventhub.Event
import com.adobe.marketing.mobile.internal.eventhub.Log
import kotlinx.coroutines.delay

internal class ExtensionA(extensionApi: SuspendableExtensionApi) : ExtensionV2(extensionApi) {
    override suspend fun readyForEvent(event: Event): Boolean {
        return true;
    }

    override suspend fun onRegistered() {
        getApi().registerEventListener("",""){ event ->
            Log.print("ExtensionA received event: " + event.getNumber())
        }
        Log.print("ExtensionA - onRegistered is done.")
    }
}

internal class ExtensionB(extensionApi: SuspendableExtensionApi) : ExtensionV2(extensionApi) {
    override suspend fun readyForEvent(event: Event): Boolean {
        return true;
    }

    override suspend fun onRegistered() {
        getApi().registerEventListener("",""){ event ->
            Log.print("ExtensionB received event: " + event.getNumber())
            NonBlockingOperation.sendRequest(event.getNumber())
            Log.print("ExtensionB processed event: " + event.getNumber())
        }
        Thread.sleep(1000)
        Log.print("ExtensionB - onRegistered (block 1 second) is done.")
    }
}

internal class ExtensionC(extensionApi: SuspendableExtensionApi) : ExtensionV2(extensionApi) {
    override suspend fun readyForEvent(event: Event): Boolean {
        val eventNumberLessThanFive = getApi().getSharedState(event.getNumber())
        if(eventNumberLessThanFive == null){
            return false
        }else {
            return true;
        }
    }

    override suspend fun onRegistered() {
        getApi().registerEventListener("",""){ event ->
            Log.print("ExtensionC received event: " + event.getNumber())
        }
        Log.print("ExtensionC - onRegistered is done.")
    }
}


internal object NonBlockingOperation {
    private suspend fun saveHit(hit: Int) {
        when(hit){
            1 -> delay(2000)
            2 -> delay(1000)
            3 -> delay(500)
            4 -> delay(500)
            5 -> delay(500)
        }
    }

    suspend fun sendRequest(hit: Int): Any {
        saveHit(hit)
        // send request
        delay(100)
        return {}
    }
}